src/lsp.ts

```diff
@@ -0,0 +1,180 @@
+import * as Admzip from "adm-zip";
+import * as which from "which";
+import * as https from "https";
+import * as crypto from "crypto";
+import * as fs from "fs";
+import * as os from "os";
+import * as path from "path";
+import * as vscode from "vscode";
+import { ExtensionContext, WorkspaceConfiguration, workspace } from "vscode";
+import { Executable, LanguageClient, LanguageClientOptions, ServerOptions, TransportKind } from "vscode-languageclient/node";
+
+export abstract class LanguageServerClient {
+  config: WorkspaceConfiguration;
+  ls: LanguageClient | null = null;
+  executableName: string;
+  context: ExtensionContext;
+  private clientOptions: LanguageClientOptions = {
+    documentSelector: ["meson", { "scheme": "file", language: "meson" }]
+  };
+  abstract repoURL: string;
+  abstract get debugExe(): Executable
+  abstract get runExe(): Executable
+
+  constructor(executableName: string, context: ExtensionContext) {
+    this.executableName = executableName;
+    this.context = context;
+    this.config = workspace.getConfiguration("mesonbuild");
+    const serverModule = this.languageServerPath;
+
+    if (serverModule == null)
+      return;
+    this.startLanguageServer();
+  }
+
+  abstract supportsSystem(os: string, arch: string): boolean
+
+  abstract get downloadInfo(): [url: string, hash: string] | null;
+
+  restart(): void {
+    this.dispose();
+
+    if (this.languageServerPath == null)
+      return;
+    this.startLanguageServer();
+  }
+
+  private startLanguageServer() {
+    const serverOptions: ServerOptions = {
+      run: this.runExe,
+      debug: this.debugExe,
+      transport: TransportKind.stdio
+    };
+
+    this.ls = new LanguageClient("Swift-MesonLSP", "Meson Language Server", serverOptions, this.clientOptions, true);
+    this.ls.start();
+  }
+
+  canDownloadLanguageServer(): boolean {
```

Comment:
I think the logic in this function is strange. Whether you can download the language server is different that if you have currently found it on the current system

Reply:
I have difficulties finding a name. That's the workflow I've thought of (Pseudo code):


1. If the language server is installed, just start it and do nothingmore
2. If the OS is linux and a supported distribution+version: Download script that sets up the package sources and installs it
3. If the OS is linux and is not a supported distribution: Redirect user to the browser with the instructions to install from source
4. If the OS is Windows/MacOS: Just download+extract the zip files

A fallback could be just taking a nearly statically linked Ubuntu 18.04 binary and hoping for the best